viscy/translation/evaluation.py

```diff
@@ -12,43 +15,254 @@
+                    f"Expected 2D segmentation, got {pred.shape[0]} and {target.shape[0]}"
+                )
+            elif current_mode == "3D" and not (
+                pred.shape[0] > 1 and target.shape[0] > 1
+            ):
+                raise ValueError(
+                    f"Expected 3D segmentation, got {pred.shape[0]} and {target.shape[0]}"
+                )
+
         pred = pred[0]
         target = target[0]
+
+        # Common preprocessing for both modes
         pred_binary = pred > 0
         target_binary = target > 0
+
+        if current_mode == "2D":
+            self._compute_2d_metrics(pred, target, pred_binary, target_binary, batch)
+        else:  # 3D mode
+            self._compute_3d_metrics(pred, target, pred_binary, target_binary, batch)
+
+    def _compute_2d_metrics(self, pred, target, pred_binary, target_binary, batch):
+        """Compute and log metrics for 2D segmentation."""
         coco_metrics = mean_average_precision(pred, target)
         _logger.debug(coco_metrics)
-        self.logger.log_metrics(
-            {
-                "position": batch["position_idx"][0],
-                "time": batch["time_idx"][0],
-                "accuracy": (accuracy(pred_binary, target_binary, task="binary")),
-                "dice": (
-                    dice_score(
-                        pred_binary.long()[None],
-                        target_binary.long()[None],
-                        num_classes=2,
-                        input_format="index",
-                    )
-                ),
-                "jaccard": (jaccard_index(pred_binary, target_binary, task="binary")),
-                "mAP": coco_metrics["map"],
-                "mAP_50": coco_metrics["map_50"],
-                "mAP_75": coco_metrics["map_75"],
-                "mAR_100": coco_metrics["mar_100"],
-            }
+
+        # Create metrics dictionary
+        metrics_dict = {
+            "position": batch["position_idx"][0],
+            "time": batch["time_idx"][0],
+            "accuracy": (accuracy(pred_binary, target_binary, task="binary")),
+            "dice": (
+                dice_score(
+                    pred_binary.long()[None],
+                    target_binary.long()[None],
+                    num_classes=2,
+                    input_format="index",
+                )
+            ),
+            "jaccard": (jaccard_index(pred_binary, target_binary, task="binary")),
+            "mAP": coco_metrics["map"],
+            "mAP_50": coco_metrics["map_50"],
+            "mAP_75": coco_metrics["map_75"],
+            "mAR_100": coco_metrics["mar_100"],
+        }
+
+        # Add metadata if available
+        if "cell_type" in batch:
+            metrics_dict["cell_type"] = batch["cell_type"]
+        if "organelle" in batch:
+            metrics_dict["organelle"] = batch["organelle"]
+        if "infection_condition" in batch:
+            metrics_dict["infection_condition"] = batch["infection_condition"]
```

Comment:
These should be isolated in a dynacell-specific method.

Reply:
These are also repeated below so the said method should be called in `test_step`.